guest/vmhome/systemtap/man/stap.1
.\" te       -*- nroff -*-
.TH STAP 1 
.SH NAME
stap \- systemtap script translator/driver

.\" macros
.\" do not nest SAMPLEs
.de SAMPLE
.br

.nr oldin \\n(.i
.RS
.nf
.nh
..
.de ESAMPLE
.hy
.fi
.RE
.in \\n[oldin]u

..

.SH SYNOPSIS

.br
.B stap
[
.I OPTIONS
]
.I FILENAME
[
.I ARGUMENTS
]
.br
.B stap
[
.I OPTIONS
]
.B \-
[
.I ARGUMENTS
]
.br
.B stap
[
.I OPTIONS
]
.BI \-e " SCRIPT"
[
.I ARGUMENTS
]
.br
.B stap
[
.I OPTIONS
]
.BI \-l " PROBE"
[
.I ARGUMENTS
]
.br
.B stap
[
.I OPTIONS
]
.BI \-L " PROBE"
[
.I ARGUMENTS
]
.br
.B stap
[
.I OPTIONS
]
.B \-\-dump-probe-types
.br
.B stap
[
.I OPTIONS
]
.B \-\-dump-probe-aliases
.br
.B stap
[
.I OPTIONS
]
.B \-\-dump-functions



.SH DESCRIPTION

The
.IR stap
program is the front-end to the Systemtap tool.  It accepts probing
instructions written in a simple domain-specific language, translates
those instructions into C code, compiles this C code, and loads the
resulting module into a running Linux kernel or a Dyninst user-space
mutator, to perform the requested system trace/probe functions.  You
can supply the script in a named file (FILENAME), from standard input
(use \- instead of FILENAME), or from the command line (using \-e
SCRIPT).  The program runs until it is interrupted by the user, or if
the script voluntarily invokes the
.I exit()
function, or by sufficient number of soft errors.
.PP
The language, which is described the
.IR "SCRIPT LANGUAGE" 
section below, is strictly typed, expressive, declaration free, procedural,
prototyping-friendly, and inspired by
.IR awk " and " C .
It allows source code points or events in the system to be associated
with handlers, which are subroutines that are executed synchronously.  It is
somewhat similar conceptually to "breakpoint command lists" in the
.IR gdb
debugger.

.SH "DOCUMENTATION OVERVIEW"

systemtap comes with a variety of educational, documentation and
reference resources.  They come online and/or packaged for offline
use.  Some systemtap diagnostic warning/error messages specially
suggest reading a man page by including a string
like \%[man\ error::pass5].  For online documentation, see the project web site,
.nh
https://sourceware.org/systemtap/
.hy
.

.TS
tab(%) allbox;
l l.
\fBman pages\fP
stap (this page)%language syntax, concepts, operation, options
error::*%further explanation of error conditions
warning::*%further explanation of warning conditions
stapprobes%probe points and their $context variables
stapref%quick reference to language syntax
stappaths%T{
list of directories, including books & references
T}
stap-prep%T{
program to install auxiliary dependencies like kernel debuginfo
T}
tapset::*%generated list of tapsets
probe::*%generated list of tapset probe aliases
function::*%generated list of tapset functions
macro::*%generated list of tapset macros
stapvars%some of the tapset global variables
staprun, stapdyn, stapbpf%programs for executing compiled systemtap scripts
systemtap%initscript, boot-time probing
stap-server%compilation server
stapex%a few very basic script examples
\fBbooks\fP%
Beginner's Guide%tutorial book, language essentials, examples
Tutorial%shorter tutorial, exercises
Language Reference%detailed language manual, covers statistics/analysis
Tapset Reference%the tapset man pages, reformatted into a book
\fBreferences\fP%
example scripts%T{
over a hundred directly usable sysadmin tools, toys, hacks to learn from
T}
.TE

.SH OPTIONS
The systemtap translator supports the following options.  Any other option
prints a list of supported options.  Options may be given on the command line,
as usual.  If the file $SYSTEMTAP_DIR/rc exist, options are also loaded from
there and interpreted first.  ($SYSTEMTAP_DIR defaults to $HOME/.systemtap if unset.)

.PP
In some cases, the default value of an option depends on particular system
configuration and thus can't be mentioned here directly.  In some of those
cases running "stap \-\-help" might display the default.

.TP
.B \-
Use standard input instead of a given FILENAME as probe language input,
unless \-e SCRIPT is given.
.TP
.B \-h \-\-help
Show help message.
.TP
.B \-V \-\-version
Show version message.
.TP
.BI \-p " NUM"
Stop after pass NUM.  The passes are numbered 1-5: parse, elaborate,
translate, compile, run.  See the
.B PROCESSING
section for details.
.TP
.B \-v
Increase verbosity for all passes.  Produce a larger volume of
informative (?) output each time option repeated.
.TP
.B \-\-vp ABCDE
Increase verbosity on a per-pass basis.  For example, "\-\-vp\ 002"
adds 2 units of verbosity to pass 3 only.  The combination "\-v\ \-\-vp\ 00004"
adds 1 unit of verbosity for all passes, and 4 more for pass 5.
.TP
.B \-k
Keep the temporary directory after all processing.  This may be useful
in order to examine the generated C code, or to reuse the compiled
kernel object.
.TP
.B \-g
Guru mode.  Enable parsing of unsafe expert-level constructs like
embedded C.
.TP
.B \-P
Prologue-searching mode.  This is equivalent to \-\-prologue\-searching=always.
Activate heuristics to work around incorrect debugging information for
function parameter $context variables.
.TP
.B \-u
Unoptimized mode.  Disable unused code elision and many other optimizations
during elaboration / translation.
.TP
.B \-w
Suppressed warnings mode.  Disables all warning messages.
.TP
.B \-W
Treat all warnings as errors.
.TP
.BI \-b
Use bulk mode (percpu files) for kernel-to-user data transfer.  Use the
.IR stap\-merge
program to multiplex them back together later.
.TP
.B \-i \-\-interactive
Interactive mode. Enable an interface to build the systemtap script
incrementally and interactively.
.TP
.B \-t
Collect timing information on the number of times probe executes
and average amount of time spent in each probe-point. Also shows 
the derivation for each probe-point.
.TP
.BI \-s " NUM"
Use NUM megabyte buffers for kernel-to-user data transfer per processor.
The default is 16MB, or less on smaller memory machines.
.TP
.BI \-I " DIR"
Add the given directory to the tapset search directory.  See the
description of pass 2 for details.
.TP
.BI \-D " NAME=VALUE"
Add the given C preprocessor directive to the module Makefile.  These can
be used to override limit parameters described below.
.TP
.BI \-B " NAME=VALUE"
In kernel-runtime mode, add the given make directive to the kernel module
build's make invocation.  These can be used to add or override kconfig options.
For example, use
.SAMPLE
.B \-B CONFIG_DEBUG_INFO=y
.ESAMPLE
to add debugging information.
.TP
.BI \-B " FLAG"
In dyninst-runtime mode, add the given parameter to the compiler CFLAGS used
for building the dyninst shared library.  For example, use
.SAMPLE
.B \-B \-g
.ESAMPLE
to add debugging information.
.TP
.BI \-a " ARCH"
Use a cross-compilation mode for the given target architecture.  This requires
access to the cross-compiler and the kernel build tree, and goes along
with the
.SAMPLE
.BR "\-B CROSS_COMPILE=arch\-tool\-prefix\-"
and
.BR "\-r /build/tree"
.ESAMPLE
options.
.TP
.BI \-\-modinfo " NAME=VALUE"
Add the name/value pair as a MODULE_INFO macro call to the generated module.
This may be useful to inform or override various module-related checks in
the kernel.
.TP
.BI \-G " NAME=VALUE"
Sets the value of global variable NAME to VALUE when staprun is invoked.
This applies to scalar variables declared global in the script/tapset.
.TP
.BI \-R " DIR"
Look for the systemtap runtime sources in the given directory.
Your DIR default can be seen using "stap \-\-help".
.TP
.BI \-r " /DIR"
Build for kernel in given build tree. Can also be set with the
.I SYSTEMTAP_RELEASE
environment variable.
.TP
.BI \-r " RELEASE"
Build for kernel in build tree
.BR /lib/modules/RELEASE/build . 
Can also be set with the
.I SYSTEMTAP_RELEASE
environment variable.
.TP
.BI \-m " MODULE"
Use the given name for the generated kernel object module, instead
of a unique randomized name.  The generated kernel object module is
copied to the current directory.
.TP
.BI \-d " MODULE"
Add symbol/unwind information for the given module into the kernel object
module.  This may enable symbolic tracebacks from those modules/programs,
even if they do not have an explicit probe placed into them.
.TP
.BI \-\-ldd
Add symbol/unwind information for all user-space shared libraries suspected
by ldd to be necessary for user-space binaries being probed or listed with
the \-d option.  Caution: this can make the probe modules considerably
larger.  Note that this option does not deal with kernel-space modules:
see instead \-\-all\-modules below.
.TP
.BI \-\-all\-modules
Equivalent to specifying "\-dkernel" and a "\-d" for each kernel module that is
currently loaded.  Caution: this can make the probe modules considerably
larger.
.TP
.BI \-o " FILE"
Send standard output to named file. In bulk mode, percpu files will
start with FILE_ (FILE_cpu with \-F) followed by the cpu number.
This supports strftime(3) formats for FILE.
.TP
.BI \-c " CMD"
Start the probes, run CMD, and exit when CMD finishes.  This also has
the effect of setting target() to the pid of that process.  Note that
many probe types trigger independently of this setting.  Consider
including something like this to focus your script.
.SAMPLE
   probe FOO { if (pid() != target()) next; .... }
.ESAMPLE
.TP
.BI \-x " PID"
Sets target() to PID.  The script runs independently of the PID's lifespan.
.TP
.BI \-e " SCRIPT"
Run the given SCRIPT specified on the command line.
.TP
.BI \-E " SCRIPT"
Run the given SCRIPT specified. This SCRIPT is run in addition to the main
script specified, through -e, or as a script file. This option can be repeated
to run multiple scripts, and can be used in listing mode (\-l/\-L).
.TP
.BI \-l " PROBE"
Instead of running a probe script, just list all available probe
points matching the given single probe point.  The pattern may include
wildcards and aliases, but not comma-separated multiple probe points.
The process result code will indicate failure if there are no matches.
.SAMPLE
% stap -e 'probe syscall.* { }'
[...]
% stap -l 'syscall.*'
syscall.accept
[...]
syscall.writev
.ESAMPLE
.TP
.BI \-L " PROBE"
Similar to "\-l", but list matching probe points plus their available context variables.
When -v is set with -L, the output includes duplicate probe points which are distinguished 
by their PC address.
.SAMPLE
% stap -L 'process("/lib64/libpython*.so.*").mark("*")'
process("/usr/lib64/libpython2.7.so.1.0").mark("function__entry") $arg1:long $arg2:long $arg3:long
process("/usr/lib64/libpython2.7.so.1.0").mark("function__return") $arg1:long $arg2:long $arg3:long
process("/usr/lib64/libpython3.6m.so.1.0").mark("function__entry") $arg1:long $arg2:long $arg3:long
process("/usr/lib64/libpython3.6m.so.1.0").mark("function__return") $arg1:long $arg2:long $arg3:long
process("/usr/lib64/libpython3.6m.so.1.0").mark("gc__done") $arg1:long
process("/usr/lib64/libpython3.6m.so.1.0").mark("gc__start") $arg1:long
process("/usr/lib64/libpython3.6m.so.1.0").mark("line") $arg1:long $arg2:long $arg3:long
.ESAMPLE
.TP
.BI \-F
Without \-o option, load module and start probes, then detach from the module
leaving the probes running.
With \-o option, run staprun in background as a daemon and show its pid.
.TP
.BI \-S " size[,N]"
Sets the maximum size of output file and the maximum number of output files.
If the size of output file will exceed
.B size
megabytes, systemtap switches output file to the next file. And if the number of
output files exceed
.B N
, systemtap removes the oldest output file. You can omit the second argument.
.TP
.BI \-T " TIMEOUT"
Exit the script after TIMEOUT seconds.
.TP
.B \-\-skip\-badvars
Ignore unresolvable or run-time-inaccessible context variables and
substitute with 0, without errors.

.TP
\fB\-\-prologue\-searching\fR[=\fIWHEN\fR]
Prologue-searching mode. Activate heuristics to work around incorrect debugging
information  for  function  parameter $context variables. WHEN can be either
"never", "always", or "auto" (i.e. enabled by heuristic). If WHEN is missing,
then "always" is assumed. If the option is missing, then "auto" is assumed.

.TP
.B \-\-suppress\-handler\-errors
Wrap all probe handlers into something like this
.SAMPLE
try { ... } catch { next }
.ESAMPLE
block, which causes any runtime errors to be quietly suppressed.
Suppressed errors do not count against
.BR MAXERRORS
limits.  In this mode, the
.BR MAXSKIPPED
limits are also suppressed, so that many errors and skipped probes 
may be accumulated during a script's runtime.  Any overall counts will
still be reported at shutdown.

.TP
.BI \-\-compatible " VERSION"
Suppress recent script language or tapset changes which are incompatible
with given older version of systemtap.  This may be useful if a much older
systemtap script fails to run.  See the DEPRECATION section for more
details.

.TP
.BI \-\-check\-version
This option is used to check if the active script has any constructs
that may be systemtap version specific.  See the DEPRECATION section
for more details.

.TP
.BI \-\-clean\-cache
This option prunes stale entries from the cache directory.  This is normally
done automatically after successful runs, but this option will trigger the
cleanup manually and then exit.  See the CACHING section for more details about
cache limits.

.TP
\fB\-\-color\fR[=\fIWHEN\fR], \fB\-\-colour\fR[=\fIWHEN\fR]
This option controls coloring of error messages. WHEN can be either "never",
"always", or "auto" (i.e. enable only if at a terminal). If WHEN is missing,
then "always" is assumed. If the option is missing, then "auto" is assumed.

Colors can be modified using the SYSTEMTAP_COLORS environment variable. The
format must be of the form
.nh
\fBkey1=val1:key2=val2:key3=val3\fR ...etc.
.hy
Valid keys are
.nh
"error", "warning", "source", "caret", and "token".
.hy
Values constitute Select Graphic Rendition (SGR) parameter(s). Consult the
documentation of your terminal for the SGRs it supports. As an example, the
default colors would be expressed as
.nh
\fBerror=01;31:warning=00;33:source=00;34:caret=01:token=01\fR.
.hy
If SYSTEMTAP_COLORS is absent, the default colors will be used. If it is empty
or invalid, coloring is turned off.

.TP
.BI \-\-disable\-cache
This option disables all use of the cache directory.  No files will be either
read from or written to the cache.

.TP
.BI \-\-poison\-cache
This option treats files in the cache directory as invalid.  No files will be
read from the cache, but resulting files from this run will still be written to
the cache.  This is meant as a troubleshooting aid when stap's cached behavior
seems to be misbehaving.  If it helped, there is a probably a bug in systemtap
that the developers would like you to report.

.TP
\fB\-\-privilege\fR[=\fIstapusr\fR | =\fIstapsys\fR | =\fIstapdev\fR]
This option instructs \fIstap\fR to examine the script looking for constructs
which are not allowed for the specified privilege level (see \fIUNPRIVILEGED USERS\fR).
Compilation fails if any such
constructs are used.
If \fIstapusr\fR or \fIstapsys\fR are specified when using a compile server
(see \fI\-\-use\-server\fR),
the server will examine the script and, if compilation succeeds, the
server will cryptographically sign the resulting kernel module, certifying
that is it safe for use by users at the specified privilege level.

If \fI\-\-privilege\fR has not been specified,
\fI\-pN\fR has not been specified with N < 5,
and the invoking user is not
\fIroot\fR, and is not a member of the group \fIstapdev\fR,
then \fIstap\fR will automatically
add the appropriate \fI\-\-privilege\fR option to the options already specified.

.TP
.BI \-\-unprivileged
This option is equivalent to \fI\-\-privilege=stapusr\fR.

.TP
\fB\-\-use\-server\fR[=\fIHOSTNAME\fR[\fI:PORT\fR] | =\fIIP_ADDRESS\fR[\fI:PORT\fR] | =\fICERT_SERIAL\fR]
Specify compile\-server(s) to be used for compilation and/or in conjunction
with
.I \-\-list\-servers
and
.I \-\-trust\-servers
(see below) for listing. If no argument is
supplied, then the default in unprivileged mode (see
.IR \-\-privilege )
is to select compatible servers which are trusted as SSL peers and as
module signers and currently online. Otherwise the default is to select
compatible servers which are trusted as SSL peers
and currently online.
.I \-\-use\-server
may be
specified more than once, in which case a list of servers is accumulated
in the order specified. Servers may be specified by host name, ip address, or
by certificate serial number (obtained using
.IR \-\-list\-servers ).
The latter is most commonly used when adding or revoking
trust in a server (see
.I \-\-trust\-servers
below). If a server is specified by host name or ip address, then an optional
port number may be specified. This is useful for accessing servers which are
not on the local network or to specify a particular server.

IP addresses may be IPv4 or IPv6 addresses.

If a particular IPv6 address is link local and exists
on more than one interface, the intended interface may be specified by appending the address with
a percent sign (%) followed by the intended interface name. For example,
"fe80::5eff:35ff:fe07:55ca%eth0".

In order to specify a port number with an IPv6 address, it is necessary to enclose the IPv6 address
in square brackets ([]) in order to separate the port number from the rest of the address. For
example, "[fe80::5eff:35ff:fe07:55ca]:5000" or "[fe80::5eff:35ff:fe07:55ca%eth0]:5000".

If \fI\-\-use\-server\fR has not been specified,
\fI\-pN\fR has not been specified with N < 5,
and the invoking user not \fIroot\fR,
is not a member of the group \fIstapdev\fR, but is a member of the group
\fIstapusr\fR, then \fIstap\fR will automatically
add \fI\-\-use\-server\fR to the options already specified.

.TP
\fB\-\-use\-server\-on\-error\fR[=\fByes\fR|=\fBno\fR]
Instructs stap to retry compilation of a script using a compile server if
compilation on the local host fails in a manner which suggests that it might
succeed using a server.
If this option is not specified, the default is \fIno\fR.
If no argument is provided, then the default
is \fIyes\fR. Compilation will be retried for certain types of errors
(e.g. insufficient data or resources) which may not occur during
re\-compilation by a compile
server. Compile servers will be selected automatically for the
re\-compilation attempt as if \fI\-\-use\-server\fR was specified with no
arguments.

.TP
.BI \-\-list\-servers "[=SERVERS]"
Display the status of the requested
.IR SERVERS ,
where
.I SERVERS
is a comma\-separated
list of server attributes. The list of attributes is combined to filter the
list of servers displayed. Supported attributes are:
.RS
.TP
.BI all
specifies all known servers (trusted SSL peers, trusted module signers, online
servers).
.TP
.BI specified
specifies servers specified using
.IR \-\-use\-server .
.TP
.BI online
filters the output by retaining information about servers which are currently
online.
.TP
.BI trusted
filters the output by retaining information about servers which are trusted as
SSL peers.
.TP
.BI signer
filters the output by retaining information about servers which are trusted as
module signers (see
.IR \-\-privilege ).
.TP
.BI compatible
filters the output by retaining information about servers which are compatible
with the current kernel release and architecture.
.RE
.IP
If no argument is provided, then the default is
.BR specified . 
If no servers were specified using
.IR \-\-use\-server ,
then the default servers for
.IR \-\-use\-server
are listed.

Note that
.IR \-\-list\-servers
uses the
.IR avahi\-daemon
service to detect online servers. If this service is not available, then
.IR \-\-list\-servers
will fail to detect any
.IR online
servers. In order for
.IR \-\-list\-servers
to detect servers listening on IPv6 addresses, the
.IR avahi\-daemon
configuration file
.IR /etc/avahi/avahi-daemon.conf
must contain an active "use-ipv6=yes" line. The service must be restarted after adding this line
in order for IPv6 to be enabled.

.TP
.BI \-\-trust\-servers "[=TRUST_SPEC]"
Grant or revoke trust in compile\-servers, specified using
.IR \-\-use\-server
as specified by TRUST_SPEC,
where TRUST_SPEC is a comma\-separated list specifying the trust which is to
be granted or revoked. Supported elements are:
.RS
.TP
.BI ssl
trust the specified servers as SSL peers.
.TP
.BI signer
trust the specified servers as module signers (see
.IR \-\-privilege ).
Only root can specify
.BR signer.
.TP
.BI all\-users
grant trust as an ssl peer for all users on the local host. The default is
to grant trust as an ssl peer for the current user only. Trust as a module
signer is always granted for all users. Only root can specify
.BR all\-users .
.TP
.BI revoke
revoke the specified trust. The default is to grant it.
.TP
.BI no\-prompt
do not prompt the user for confirmation before carrying out the requested
action. The default is to prompt the user for confirmation.
.RE
.IP
If no argument is provided, then the default is
.BR ssl . 
If no servers were specified using
.IR \-\-use\-server ,
then no trust will be granted or revoked.
.IP
Unless \fBno\-prompt\fR has been specified,
the user will be prompted to confirm the trust to be granted or revoked before
the operation is performed.

.TP
.BI \-\-sign-module
Sign the module with a MOK (Machine Owner Key) on UEFI/SecureBoot systems.
See the SECUREBOOT section for more details.

.TP
.BI \-\-dump-probe-types
Dumps a list of supported probe types and exits. If
.IR \-\-privilege=stapusr
is also specified, the list will be limited to probe types available to unprivileged users.

.TP
.BI \-\-dump-probe-aliases
Dumps a list of all probe aliases found in library files and exits.

.TP
.BI \-\-dump-functions
Dumps a list of all the public functions found in library files and exits. Also includes
their parameters and types. A function of type 'unknown' indicates a function
that does not return a value. Note that not all function/parameter types may be
resolved (these are also shown by 'unknown'). This features is very
memory-intensive and thus may not work properly with \fI\-\-use-server\fR if the
target server imposes an rlimit on process memory (i.e. through the
\fI~stap-server/.systemtap/rc\fR configuration file, see \fIstap-server\fR(8)).

.TP
.BI \-\-remote " URL"
Set the execution target to the given host.  This option may be
repeated to target multiple execution targets.  Passes 1-4 are
completed locally as normal to build the script, and then pass 5 will
copy the module to the target and run it.  Acceptable URL forms include:
.RS
.TP
\fB[USER@]HOSTNAME\fR, \fBssh://[USER@]HOSTNAME\fR
This mode uses ssh, optionally using a username not matching your own. If a
custom ssh_config file is in use, add \fBSendEnv LANG\fR to retain
internationalization functionality.
.TP
\fBlibvirt://DOMAIN\fR, \fBlibvirt://DOMAIN/LIBVIRT_URI\fR
This mode uses \fIstapvirt\fR to execute the script on a domain managed by
libvirt. Optionally, LIBVIRT_URI may be specified to connect to a specific
driver and/or a remote host. For example, to connect to the local privileged
QEMU driver, use:
.SAMPLE
\-\-remote libvirt://MyDomain/qemu:///system
.ESAMPLE
See the page at
.nh
<http://libvirt.org/uri.html>
.hy
for supported URIs. Also see \fIstapvirt\fR(1) for more information on how to
prepare the domain for stap probing.
.TP
\fBunix:PATH\fR
This mode connects to a UNIX socket. This can be used with a QEMU virtio-serial
port for executing scripts inside a running virtual machine.
.TP
\fBdirect://\fR
Special loopback mode to run on the local host.
.RE
.IP

.TP
.BI \-\-remote\-prefix
Prefix each line of remote output with "N: ", where N is the index of the remote
execution target from which the given line originated.

.TP
.BI \-\-download\-debuginfo "[=OPTION]"
Enable, disable or set a timeout for the automatic debuginfo downloading feature
offered by abrt as specified by OPTION, where OPTION is one of the following:
.RS
.TP
.BI yes
enable automatic downloading of debuginfo with no timeout. This is the same
as not providing an OPTION value to 
.IR \-\-download\-debuginfo
.TP
.BI no
explicitly disable automatic downloading of debuginfo. This is the same as
not using the option at all.
.TP
.BI ask
show abrt output, and ask before continuing download. No timeout will be set.
.TP
.BI <timeout>
specify a timeout as a positive number to stop the download if it is taking 
longer than <timeout> seconds.
.RE
.IP

.TP
.BI \-\-rlimit-as "=NUM"
Specify the maximum size of the process's virtual memory (address space),
in bytes.

.TP
.BI \-\-rlimit-cpu "=NUM"
Specify the CPU time limit, in seconds.

.TP
.BI \-\-rlimit-nproc "=NUM"
Specify the maximum number of processes that can be created.

.TP
.BI \-\-rlimit-stack "=NUM"
Specify the maximum size of the process stack, in bytes.

.TP
.BI \-\-rlimit-fsize "=NUM"
Specify the maximum size of files that the process may create, in bytes.

.TP
.BI \-\-sysroot "=DIR"
Specify sysroot directory where target files (executables, libraries, etc.)
are located.  With \fI\-r RELEASE\fR, the sysroot will be searched for the
appropriate kernel build directory.  With \fI\-r /DIR\fR, however, the sysroot
will not be used to find the kernel build.

.TP
.BI \-\-sysenv "=VAR=VALUE"
Provide an alternate value for an environment variable where the value on a
remote system differs.  Path variables (e.g. PATH, LD_LIBRARY_PATH) are assumed
to be relative to the directory provided by \fI\-\-sysroot\fR, if provided.

.TP
.BI \-\-suppress\-time\-limits
Disable \-DSTP_OVERLOAD related options as well as \-DMAXACTION and \-DMAXTRYLOCK.
This option requires guru mode.

.TP
.BI \-\-runtime "=MODE"
Set the pass-5 runtime mode.  Valid options are \fIkernel\fR (default),
\fIdyninst\fR and \fIbpf\fR.  See
.BR ALTERNATE\ RUNTIMES
below for more information.

.TP
.BI \-\-dyninst
Shorthand for \fI\-\-runtime=dyninst\fR.

.TP
.BI \-\-bpf
Shorthand for \fI\-\-runtime=bpf\fR.

.TP
.BI \-\-save-uprobes
On machines that require SystemTap to build its own uprobes module (kernels
prior to version 3.5), this option instructs SystemTap to also save a copy of
the module in the current directory (creating a new "uprobes" directory first).

.TP
.BI \-\-target-namespaces "=PID"
Allow for a set of target namespaces to be set based on the namespaces the
given PID is in. This is for namespace\-aware tapset functions. If the target
namespaces was not set, the target defaults to the stap process'
namespaces.

.TP
.BI \-\-monitor "=INTERVAL"
Enables an interface to display status information about the module(uptime,
module name, invoker uid, memory sizes, global variables, list of probes with
their statistics). An optional argument INTERVAL can be supplied to set the
refresh rate in seconds of the status window. The module can also be controlled by a list
of commands using the following keys:
.RS
.TP
.BI c
Resets all global variables to their initial values or zeroes them if they
did not have an initial value.
.TP
.BI s
Rotates the attribute used to sort the list of probes.
.TP
.BI t
Brings up a prompt to allow toggling(on/off) of probes by index. Probe points
are still affected by their conditions.
.TP
.BI r
Resumes the script by toggling on all probes.
.TP
.BI p
Pauses the script by toggling off all probes.
.TP
.BI x
Hides/shows the status window. This allows for more output to be seen.
.TP
.BI navigation-keys
The navigation keys can be used to scroll up and down the windows.
.TP
.BI Tab
Toggle scrolling between status and output windows.
.RE

.TP
.BI \-\-example
This option is used to run example scripts without having to enter the
entire path to the script. Example scripts can be found in the directory
specified in the
.IR stappaths (7)
manual page.

.TP
.BI \-\-no-global-var-display
This option is used to disable the automatic logging of unused global
variables at the end of a stap session.

.TP
.B \-\-language\-server
Language server mode. Start a language server which will communicate via
stdio. The language server will respect stap verbosity.

.TP
.B \-\-build\-as
Execute passes 1-4 under specified user to increase safety.  This approach
may have limitations.  Unprivileged users might not have access to kallsyms
or /proc.  At the RPM install time, the stapunpriv user is created, so stap
can be invoked with --build-as=stapunpriv.

.SH ARGUMENTS

Any additional arguments on the command line are passed to the script
parser for substitution.  See below.

.SH SCRIPT LANGUAGE

The systemtap script language resembles
.IR awk " and " C .
There are two main outermost constructs: probes and functions.  Within
these, statements and expressions use C-like operator syntax and
precedence.

.SS GENERAL SYNTAX
Whitespace is ignored.  Three forms of comments are supported:
.RS
.br
.BR # " ... shell style, to the end of line, except for $# and @#"
.br
.BR // " ... C++ style, to the end of line"
.br
.BR /* " ... C style ... " */
.RE
Literals are either strings enclosed in double-quotes (passing through
the usual C escape codes with backslashes, and with adjacent string
literals glued together, also as in C), or integers (in decimal,
hexadecimal, or octal, using the same notation as in C).  All strings
are limited in length to some reasonable value (a few hundred bytes).
Integers are 64-bit signed quantities, although the parser also
accepts (and wraps around) values above positive 2**63.
.PP
In addition, script arguments given at the end of the command line may
be inserted.  Use
.B $1 ... $<NN>
for insertion unquoted,
.B @1 ... @<NN>
for insertion as a string literal.  The number of arguments may be accessed
through
.B $#
(as an unquoted number) or through
.B @#
(as a quoted number).  These may be used at any place a token may begin,
including within the preprocessing stage.  Reference to an argument
number beyond what was actually given is an error.

.SS PREPROCESSING
A simple conditional preprocessing stage is run as a part of parsing.
The general form is similar to the
.RB cond " ? " exp1 " : " exp2
ternary operator:
.SAMPLE
.BR %( " CONDITION " %? " TRUE-TOKENS " %)
.BR %( " CONDITION " %? " TRUE-TOKENS " %: " FALSE-TOKENS " %)
.ESAMPLE
The CONDITION is either an expression whose format is determined by its
first keyword, or a string literals comparison or a numeric literals
comparison.  It can be also composed of many alternatives and conjunctions
of CONDITIONs (meant as in previous sentence) using || and && respectively.
However, parentheses are not supported yet, so remembering that conjunction
takes precedence over alternative is important.
.PP
If the first part is the identifier
.BR kernel_vr " or " kernel_v
to refer to the kernel version number, with ("2.6.13\-1.322FC3smp") or
without ("2.6.13") the release code suffix, then
the second part is one of the six standard numeric comparison operators
.BR < ", " <= ", " == ", " != ", " > ", and " >= ,
and the third part is a string literal that contains an RPM-style
version-release value.  The condition is deemed satisfied if the
version of the target kernel (as optionally overridden by the
.BR \-r
option) compares to the given version string.  The comparison is
performed by the glibc function
.BR strverscmp .
As a special case, if the operator is for simple equality
.RB ( == ),
or inequality
.RB ( != ),
and the third part contains any wildcard characters
.RB ( * " or " ? " or " [ "),"
then the expression is treated as a wildcard (mis)match as evaluated
by
.BR fnmatch .
.PP
If, on the other hand, the first part is the identifier
.BR arch
to refer to the processor architecture (as named by the kernel
build system ARCH/SUBARCH), then the second  
part is one of the two string comparison operators
.BR == " or " != ,
and the third part is a string literal for matching it.  This
comparison is a wildcard (mis)match.
.PP
Similarly, if the first part is an identifier like
.BR CONFIG_something
to refer to a kernel configuration option, then the second part is
.BR == " or " != ,
and the third part is a string literal for matching the value
(commonly "y" or "m").  Nonexistent or unset kernel configuration
options are represented by the empty string.  This comparison is also
a wildcard (mis)match.
.PP
If the first part is the identifier
.BR systemtap_v ,

guest/vmhome/systemtap/interactive-notebook/codemirror/lib/stap.js
// The stap highlighting rules
// Copyright (C) 2023 Red Hat Inc.
//
// This file is part of systemtap, and is free software.  You can
// redistribute it and/or modify it under the terms of the GNU General
// Public License (GPL); either version 2, or (at your option) any
// later version.
import * as CodeMirror from 'codemirror';
import 'codemirror/addon/mode/simple';

CodeMirror.defineSimpleMode("systemtap", {
  // The start state contains the rules that are intially used
  // Rules are matched in order. I am using the same order as vscode/atom
  start: [
    // comments
    { regex: /\/\/\/?.*$/, token: 'comment' },
    { regex: /#.*$/, token: "comment" },
    { regex: /\/\*/, token: 'comment', push: 'comment_block' },
    
    // embedded C
    { regex: /\%{/, token: "meta", mode: { spec: "clike", end: /\%}/}},

    // numbers
    // octal number
    { regex: /(0)([0-7]+)/, token: ["string", "number"] },
    // hex number
    { regex: /(0x)([0-9a-fA-F]+)/, token: ["string", "number"] },
    // numeric args
    { regex: /\$#/, token: "number" },
    { regex: /\$\d+/, token: "number" },
    // decimal number
    { regex: /\d+/, token: "number" },

    // string
    { regex: /"/, token: 'string', push: 'string' },
    
    // TODO: probe signatures could be highlighted
    {
      regex: /probe/,
      token: "keyword",
    },

    // functions (1 group per token)
    {
      regex: /(function)(\s+)([A-Za-z\$_][0-9A-Za-z_\$]*)/,
      token: ["keyword", null, "variable-2"],
    },

    // keywords
    // Should I add (?<!\.) I can't let it start with . since .return probes exist
    {
      regex: /(\b)(break|continue|return|next|delete|try|catch|while|for|foreach|in|limit|if|else|global|function|private|probe)(\b)/,
      token: [null, "keyword", null]
    },

    //macros & atwords
    {
      regex: /@\w+/,
      token: "atom"
    },

    // types
    {
      regex: /(\b)(string|long)(\b)/,
      token: "keyword"
    },

    // Magic commands
    {
      regex: /^%%.*$/,
      token: "string-2"
    },

    // variables
    {
      regex: /\b[A-Za-z\$_][0-9A-Za-z_\$]*\b/,
      token: "variable"
    },

    // operators
    {
      regex: /[\*\/\%\+\-\&\^!=<>~|.]=?|[=!~:?]/,
      token: "operator"
    },

    {
      regex: /\{%\{%\(/,
      indent: true
    },
    {
      regex: /\}%\}%\)/,
      dedent: true
    },

    // punctuation
    { regex: /(\;|\||,)/, token: "punctuation" }
  ],

  comment_block: [
    { regex: /\/\*/, token: 'comment', push: 'comment_block' },
    // this ends and restarts a comment block. but need to catch this so
    // that it doesn\'t start _another_ level of comment blocks
    { regex: /\*\/\*/, token: 'comment' },
    { regex: /(\*\/\s+\*(?!\/)[^\n]*)|(\*\/)/, token: 'comment', pop: true },
    // Match anything else as a character inside the comment
    { regex: /./, token: 'comment' },
  ],

  string: [
    { regex: /"/, token: 'string', pop: true },
    { regex: /./, token: 'string' }
  ],

  // probepoints: [
  //   { regex: /0[0-7]+/, token: "number" },
  //   { regex: /0x[0-9a-fA-F]+/, token: "number" },
  //   { regex: /\d+/, token: "number" },
  //   { regex: /"/, token: 'string', push: 'string' },
  //   // (?={) looksahead for { 
  //   { regex: /.*(?={)/, token: 'comment', pop: true },
  // ],

  meta: {
    closeBrackets: { pairs: "()[]\{\}\"\"%\{\}%(%)" },
    dontIndentStates: ['comment_block'],
    electricInput: /^\s*\}$/,
    blockCommentStart: '/*',
    blockCommentEnd: '*/',
    lineComment: '//',
    fold: 'brace'
  }
});

CodeMirror.defineMIME('text/x-systemtap', 'systemtap');
CodeMirror.defineMIME('text/systemtap', 'systemtap');

CodeMirror.modeInfo.push({
  ext: ['stp'],
  mime: "text/x-systemtap",
  mode: 'systemtap',
  name: 'systemtap'
});
